Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,405 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseFirestore
+@testable import FirebaseFirestoreSwift
+import XCTest
+
+class CodableDocumentTests: XCTestCase {
+  func roundTrip<X>(input: X, expected: [String: Any]? = nil) -> X where X: Codable {
+    var encoded = [String: Any]()
+    do {
+      encoded = try Firestore.encode(input)
+      if let expected = expected {
+        XCTAssertEqual(encoded as NSDictionary, expected as NSDictionary)
+      }
+    } catch {
+      XCTFail("Failed to encode \(X.self): error: \(error)")
+    }
+    do {
+      let decoded = try Firestore.decode(X.self, from: encoded)
+      return decoded
+    } catch {
+      XCTFail("Failed to decode \(X.self): \(error)")
+    }
+    return input // After failure
+  }
+
+  func testInt() {
+    struct Model: Codable {
+      let x: Int
+    }
+    let model = Model(x: 42)
+    let dict = ["x": 42]
+    XCTAssertEqual(model.x, roundTrip(input: model, expected: dict).x)
+  }
+
+  func testEmpty() {
+    struct Model: Codable {}
+    let model = Model()
+    let dict = [String: Any]() as NSDictionary
+    let encoded = try! Firestore.encode(model) as NSDictionary
+    XCTAssertEqual(encoded, dict)
+  }
+
+  func testNil() {
+    struct Model: Codable {
+      let x: Int?
+    }
+    let model = Model(x: nil)
+    let dict = ["x": nil] as [String: Int?]
+    let encodedDict = try! Firestore.encode(model)
+    XCTAssertNil(encodedDict["x"])
+    let model2 = try? Firestore.decode(Model.self, from: dict as [String: Any])
+    XCTAssertNil(model2)
+  }
+
+  func testOptional() {
+    struct Model: Codable {
+      let x: Int
+      let opt: Int?
+    }
+    let dict = ["x": 42]
+    let model = Model(x: 42, opt: nil)
+    XCTAssertEqual(model.x, roundTrip(input: model, expected: dict).x)
+
+    let model2 = Model(x: 42, opt: 7)
+    let expected = ["x": 42, "opt": 7]
+    let encoded = try! Firestore.encode(model2)
+    XCTAssertEqual(encoded as NSDictionary, expected as NSDictionary)
+    let decoded = try! Firestore.decode(Model.self, from: expected)
+    XCTAssertEqual(decoded.x, model2.x)
+    XCTAssertEqual(decoded.opt, model2.opt)
+  }
+
+  func testOptionalTimestamp() {
+    struct Model: Codable {
+      let value: Int
+      let timestamp: Timestamp?
+    }
+    class FirestoreDummy {
+      var visited = 0
+      func setObject<T: Codable>(_ object: T, fieldValues: [PartialKeyPath<T>: FieldValue] = [:]) {
```

Comment:
I can't understand what this test is doing at all. It needs some comments describing what's going on here.

For example, is `setObject` some hook that's part of `Codable` generally? Are we simulating something an end user might write? Or are we using this to assert the behavior of the encoder? It seems like if the latter we can just  assert that a completed object serializes one way or another.

In any case, it appears that `fieldValues` is completely ignored here which doesn't seem right.

Also, asserting the visited count doesn't seem useful. It seems as if it's asserting that the test itself called setObject twice?

Reply:
Deleted the test.